File: protocolarc/loader.py
"""Loaders that turn JSON/JSONL documents into model objects.

Contract documents and envelope documents are both plain JSON. The loader is
tolerant of ordering but strict about shape: a malformed document raises
``ContractError`` with a precise message rather than silently degrading.
"""

from __future__ import annotations

import json
import os
from typing import Any, Dict, List

from .model import Contract, Envelope, FieldSpec, ContractError, JSON_TYPES


def _require(doc: Dict[str, Any], key: str, where: str) -> Any:
    if key not in doc:
        raise ContractError(f"{where}: missing required key '{key}'")
    return doc[key]


def _as_type_tuple(raw: Any, where: str):
    if isinstance(raw, str):
        raw = [raw]
    if not isinstance(raw, list) or not raw:
        raise ContractError(f"{where}: 'type' must be a string or non-empty list")
    types = tuple(str(t) for t in raw)
    for t in types:
        if t not in JSON_TYPES:
            raise ContractError(
                f"{where}: unknown type '{t}'; expected one of {', '.join(JSON_TYPES)}"
            )
    return types


def parse_contract(doc: Dict[str, Any], source: str = "<inline>") -> Contract:
    """Build a Contract from a decoded contract document."""

    if not isinstance(doc, dict):
        raise ContractError(f"{source}: contract must be a JSON object")

    name = _require(doc, "name", source)
    revision = _require(doc, "revision", source)
    fields_raw = doc.get("fields", [])
    if not isinstance(fields_raw, list):
        raise ContractError(f"{source}: 'fields' must be a list")

    fields: List[FieldSpec] = []
    seen = set()
    for idx, fraw in enumerate(fields_raw):
        where = f"{source}: fields[{idx}]"
        if not isinstance(fraw, dict):
            raise ContractError(f"{where}: each field must be an object")
        fname = _require(fraw, "name", where)
        if fname in seen:
            raise ContractError(f"{where}: duplicate field name '{fname}'")
        seen.add(fname)
        types = _as_type_tuple(_require(fraw, "type", where), where)
        enum_raw = fraw.get("enum")
        enum = tuple(enum_raw) if isinstance(enum_raw, list) else None
        fields.append(
            FieldSpec(
                name=str(fname),
                types=types,
                required=bool(fraw.get("required", True)),
                enum=enum,
                doc=str(fraw.get("doc", "")),
            )
        )

    return Contract(
        name=str(name),
        revision=str(revision),
        kind=str(doc.get("kind", "generic")),
        title=str(doc.get("title", name)),
        fields=fields,
        strict=bool(doc.get("strict", False)),
    )


def parse_envelope(doc: Dict[str, Any], source: str = "<inline>") -> Envelope:
    """Build an Envelope from a decoded envelope document.

    Two shapes are accepted:

    * *Wrapped* — an object with ``contract``, optional ``revision`` and a
      ``data`` body. This is the canonical form.
    * *Bare* — any object that carries a top-level ``contract`` key; the whole
      object (minus the routing keys) becomes the data body.
    """

    if not isinstance(doc, dict):
        raise ContractError(f"{source}: envelope must be a JSON object")

    contract = _require(doc, "contract", source)
    revision = doc.get("revision")

    if "data" in doc and isinstance(doc["data"], dict):
        data = doc["data"]
    else:
        data = {k: v for k, v in doc.items() if k not in ("contract", "revision")}

    return Envelope(
        contract=str(contract),
        revision=str(revision) if revision is not None else None,
        data=data,
        source=source,
    )


def load_contract(path: str) -> Contract:
    """Load a single contract from a JSON file."""

    with open(path, "r", encoding="utf-8") as handle:
        doc = json.load(handle)
    return parse_contract(doc, source=os.path.basename(path))


def load_contract_dir(path: str) -> List[Contract]:
    """Load every ``*.json`` contract in a directory, sorted by file name."""

    contracts: List[Contract] = []
    for entry in sorted(os.listdir(path)):
        if entry.endswith(".json"):
            contracts.append(load_contract(os.path.join(path, entry)))
    return contracts


def load_envelope(path: str) -> List[Envelope]:
    """Load envelopes from a ``.json`` (single/array) or ``.jsonl`` file."""

    base = os.path.basename(path)
    envelopes: List[Envelope] = []
    if path.endswith(".jsonl"):
        with open(path, "r", encoding="utf-8") as handle:
            for lineno, line in enumerate(handle, start=1):
                line = line.strip()
                if not line:
                    continue
                doc = json.loads(line)
                envelopes.append(parse_envelope(doc, source=f"{base}:{lineno}"))
        return envelopes

    with open(path, "r", encoding="utf-8") as handle:
        doc = json.load(handle)
    if isinstance(doc, list):
        for idx, item in enumerate(doc):
            envelopes.append(parse_envelope(item, source=f"{base}[{idx}]"))
    else:
        envelopes.append(parse_envelope(doc, source=base))
    return envelopes

File: runtime/Loader.cs
using System.Text.Json;

namespace ProtocolArc.Runtime;

/// <summary>
/// Parses contract and envelope documents from JSON using only
/// <c>System.Text.Json</c>. Malformed documents raise
/// <see cref="ContractException"/> with a precise message.
/// </summary>
public static class Loader
{
    public static Contract ParseContract(JsonElement doc, string source)
    {
        if (doc.ValueKind != JsonValueKind.Object)
            throw new ContractException($"{source}: contract must be a JSON object");

        var name = RequireString(doc, "name", source);
        var revision = RequireString(doc, "revision", source);

        var fields = new List<FieldSpec>();
        var seen = new HashSet<string>();
        if (doc.TryGetProperty("fields", out var fieldsEl) && fieldsEl.ValueKind == JsonValueKind.Array)
        {
            int idx = 0;
            foreach (var f in fieldsEl.EnumerateArray())
            {
                var where = $"{source}: fields[{idx}]";
                if (f.ValueKind != JsonValueKind.Object)
                    throw new ContractException($"{where}: each field must be an object");
                var fname = RequireString(f, "name", where);
                if (!seen.Add(fname))
                    throw new ContractException($"{where}: duplicate field name '{fname}'");

                var types = ReadTypes(f, where);
                List<JsonElement>? enumValues = null;
                if (f.TryGetProperty("enum", out var enumEl) && enumEl.ValueKind == JsonValueKind.Array)
                    enumValues = enumEl.EnumerateArray().Select(e => e.Clone()).ToList();

                bool required = !f.TryGetProperty("required", out var reqEl) ||
                                reqEl.ValueKind != JsonValueKind.False;

                string docText = f.TryGetProperty("doc", out var docEl) &&
                                 docEl.ValueKind == JsonValueKind.String
                    ? docEl.GetString() ?? ""
                    : "";

                fields.Add(new FieldSpec(fname, types, required, enumValues, docText));
                idx++;
            }
        }

        return new Contract
        {
            Name = name,
            Revision = revision,
            Kind = ReadString(doc, "kind", "generic"),
            Title = ReadString(doc, "title", name),
            Strict = doc.TryGetProperty("strict", out var s) && s.ValueKind == JsonValueKind.True,
            Fields = fields
        };
    }

    public static Envelope ParseEnvelope(JsonElement doc, string source)
    {
        if (doc.ValueKind != JsonValueKind.Object)
            throw new ContractException($"{source}: envelope must be a JSON object");

        var contract = RequireString(doc, "contract", source);
        string? revision = doc.TryGetProperty("revision", out var revEl) &&
                           revEl.ValueKind == JsonValueKind.String
            ? revEl.GetString()
            : null;

        JsonElement data;
        if (doc.TryGetProperty("data", out var dataEl) && dataEl.ValueKind == JsonValueKind.Object)
        {
            data = dataEl.Clone();
        }
        else
        {
            // Bare form: build an object without the routing keys.
            using var stream = new MemoryStream();
            using (var writer = new Utf8JsonWriter(stream))
            {
                writer.WriteStartObject();
                foreach (var prop in doc.EnumerateObject())
                {
                    if (prop.Name is "contract" or "revision") continue;
                    prop.WriteTo(writer);
                }
                writer.WriteEndObject();
            }
            stream.Position = 0;
            using var parsed = JsonDocument.Parse(stream);
            data = parsed.RootElement.Clone();
        }

        return new Envelope
        {
            Contract = contract,
            Revision = revision,
            Data = data,
            Source = source
        };
    }

    private static IReadOnlyList<string> ReadTypes(JsonElement f, string where)
    {
        if (!f.TryGetProperty("type", out var typeEl))
            throw new ContractException($"{where}: missing required key 'type'");

        var types = new List<string>();
        if (typeEl.ValueKind == JsonValueKind.String)
        {
            types.Add(typeEl.GetString()!);
        }
        else if (typeEl.ValueKind == JsonValueKind.Array)
        {
            foreach (var t in typeEl.EnumerateArray())
                types.Add(t.GetString() ?? "");
        }
        else
        {
            throw new ContractException($"{where}: 'type' must be a string or list");
        }

        foreach (var t in types)
            if (!JsonTypes.All.Contains(t))
                throw new ContractException($"{where}: unknown type '{t}'");

        if (types.Count == 0)
            throw new ContractException($"{where}: 'type' must be non-empty");
        return types;
    }

    private static string RequireString(JsonElement doc, string key, string where)
    {
        if (!doc.TryGetProperty(key, out var el))
            throw new ContractException($"{where}: missing required key '{key}'");
        return el.ValueKind == JsonValueKind.String ? el.GetString()! : el.ToString();
    }

    private static string ReadString(JsonElement doc, string key, string fallback)
    {
        return doc.TryGetProperty(key, out var el) && el.ValueKind == JsonValueKind.String
            ? el.GetString() ?? fallback
            : fallback;
    }
}

/// <summary>Raised when a document cannot be understood at all.</summary>
public sealed class ContractException : Exception
{
    public ContractException(string message) : base(message) { }
}

# draft note 52
